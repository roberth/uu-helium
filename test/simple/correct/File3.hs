module File3 where

main :: IO ()
main
  = do s1 <- readFile "File3.hs" 
       writeFile "File3.temp" s1
       s2 <- readFile "File3.temp"
       if (eqString s1 s2)
        then putStrLn "The read and writes are equal!"
        else putStrLn "*** ERROR: The read and writes are not equal?"

writeFile fname s
  = bracket (openFile fname WriteMode)
            (hClose)
            (\h -> hPutString h s)

readFile fname
  = bracket (openFile fname ReadMode)
            (hClose)
            (\h -> readAll h [])
  where
    readAll h acc 
      = do c  <- hGetChar h
           readAll h (c:acc) 
        `catchEof` (return (reverse acc))

bracket :: IO a -> (a -> IO b) -> (a -> IO c) -> IO c
bracket acquire release action
  = do x <- acquire
       finally (action x) (release x)

finally :: IO a -> IO b -> IO a
finally io action
  = do x <- io `catch` (\exn -> do{ action; raise exn })
       action
       return x
