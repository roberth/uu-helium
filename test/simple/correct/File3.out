Compiling correct/File3.hs
(12,1): Warning: Missing type signature: writeFile :: String -> String -> IO ()
(17,1): Warning: Missing type signature: readFile :: String -> IO [Char]
Compilation successful with 2 warnings
The read and writes are equal!
